### PR TITLE
docs: sqlserver data integration

### PR DESCRIPTION
**Pull request by JimMoen**

### CODE REVIEW

File: dir.yaml [hunk 1]
```diff
@@ -186,7 +186,7 @@
     # - data-integration/data-bridge-cassa
     - data-integration/data-bridge-tdengine
     - data-integration/data-bridge-rocketmq
-    
+    - data-integration/data-bridge-sqlserver
```
Comment (hjianbo) [inline dir.yaml:196]: Do we need to comment on it until e5.0.3 is released? @LenaLenaPan 
```suggestion
    #- data-integration/data-bridge-sqlserver
```

File: zh_CN/data-integration/data-bridge-sqlserver.md [hunk 2]
```diff
@@ -91,27 +91,16 @@ $ Password:
    ```
 
 
-### 创建 SQL Server 数据桥接
+### 配置 ODBC 驱动
 
-本节我们将创建一个 SQL Server 数据桥接来实现对客户端发布消息的存储。
+在开始其他步骤之前，需要您配置 ODBC 驱动。
+- 您可以使用 freetds 或 Microsoft 发布的 msodbcsql17 作为 odbc 驱动。( msodbcsql18 的连接属性并仍未进行适配 )
```
Comment (JimMoen) [inline zh_CN/data-integration/data-bridge-sqlserver.md]: FreeTDS

File: en_US/data-integration/data-bridge-sqlserver.md [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+
+1. Install SQL Server via Docker, and then run the docker image.
+
+   SQL Server requires using complex passwords, see also [Password Complexity](https://learn.microsoft.com/en-us/sql/relational-databases/security/password-policy?view=sql-server-ver16#password-complexity).
+   By starting a Docker container with the environment variable `ACCEPT_EULA=Y` you agree to the terms of Microsoft EULA, see also [MICROSOFT SOFTWARE LICENSE TERMS MICROSOFT SQL SERVER 2019 STANDARD(EN_US)](https://www.microsoft.com/en-us/Useterms/Retail/SQLServerStandard/2019/Useterms_Retail_SQLServerStandard_2019_English.htm).
+
+```bash
+# To start the SQL Server docker image and set the password as `mqtt_public1`
+$ docker run --name sqlserver -p 1433:1433 -e ACCEPT_EULA=Y -e SQLSERVER_ROOT_PASSWORD=mqtt_public1 -d mcr.microsoft.com/mssql/server:2019-CU19-ubuntu-20.04
```
Comment (JimMoen) [inline en_US/data-integration/data-bridge-sqlserver.md]: $ docker run --name sqlserver -p 1433:1433 -e ACCEPT_EULA=Y -e MSSQL_SA_PASSWORD=mqtt_public1 -d mcr.microsoft.com/mssql/server:2019-CU19-ubuntu-20.04

File: zh_CN/data-integration/data-bridge-sqlserver.md [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+本节描述如何使用 Docker 镜像在 Linux/MacOS 安装启动 SQL Server 2019 以及使用 `sqlcmd` 连接到 SQL Server 并创建数据库与数据表。关于其他 SQL Server 的安装方式，请参阅微软提供的 [SQL Server 安装指南](https://learn.microsoft.com/zh-cn/sql/database-engine/install-windows/install-sql-server?view=sql-server-ver16)。
+
+1. 通过 Docker 安装并启动 SQL Server。
+
+   SQL Server 要求使用复杂密码，请参阅 [使用复杂密码](https://learn.microsoft.com/zh-cn/sql/relational-databases/security/password-policy?view=sql-server-ver16#password-complexity)。
+   使用环境变量 `ACCEPT_EULA=Y` 启动 Docker 容器代表您同意 Microsoft 的 EULA 条款，详情请参阅 [MICROSOFT 软件许可条款 MICROSOFT SQL SERVER 2019 STANDARD(ZH_CN)](https://www.microsoft.com/en-us/Useterms/Retail/SQLServerStandard/2019/Useterms_Retail_SQLServerStandard_2019_ChineseSimplified.htm)。
+
+```bash
+# 启动一个 SQL Server 容器并设置密码为 `mqtt_public1`
+$ docker run --name sqlserver -p 1433:1433 -e ACCEPT_EULA=Y -e SQLSERVER_ROOT_PASSWORD=mqtt_public1 -d mcr.microsoft.com/mssql/server:2019-CU19-ubuntu-20.04
```
Comment (JimMoen) [inline zh_CN/data-integration/data-bridge-sqlserver.md]: $ docker run --name sqlserver -p 1433:1433 -e ACCEPT_EULA=Y -e MSSQL_SA_PASSWORD=mqtt_public1 -d mcr.microsoft.com/mssql/server:2019-CU19-ubuntu-20.04

File: en_US/data-integration/data-bridge-sqlserver.md [hunk 5]
```diff
@@ -0,0 +1,265 @@
+# Ingest Data into SQL Server
```
Comment (LenaLenaPan) [inline en_US/data-integration/data-bridge-sqlserver.md]:
```suggestion
# Ingest Data into Microsoft SQL Server
```

File: dir.yaml [hunk 6]
```diff
@@ -193,7 +193,7 @@
     # - data-integration/data-bridge-cassa
     - data-integration/data-bridge-tdengine
     - data-integration/data-bridge-rocketmq
-    
+    #- data-integration/data-bridge-sqlserver
```
Comment (LenaLenaPan) [inline dir.yaml]:
```suggestion
    - data-integration/data-bridge-sqlserver
```

File: en_US/data-integration/data-bridge-sqlserver.md [hunk 7]
```diff
@@ -0,0 +1,337 @@
...
+
+This tutorial assumes that you run both EMQX and Microsoft SQL Server on the local machine. If you have Microsoft SQL Server and EMQX running remotely, adjust the settings accordingly.
+
+### Install and Connect to Microsoft SQL Server
+
+This section describes how to start Microsoft SQL Server 2019 on Linux/MacOS using Docker images and use `sqlcmd` to connect to SQL Server. For other installation methods of SQL Server, please refer to [SQL Server Installation Guide](https://learn.microsoft.com/en-us/sql/database-engine/install-windows/install-sql-server?view=sql-server-ver16).
+
+1. Install SQL Server via Docker, and then run the docker image.
+
+   SQL Server requires using complex passwords, see also [Password Complexity](https://learn.microsoft.com/en-us/sql/relational-databases/security/password-policy?view=sql-server-ver16#password-complexity).
```
Comment (LenaLenaPan) [inline en_US/data-integration/data-bridge-sqlserver.md]: 这个解释看着有点奇怪，建议根据下面的命令调整下，Start Microsoft SQL Server 2019 with the command below and use `mqtt_public1` as the password. For the password policy of Microsoft SQL Server, see [Password Complexity](https://learn.microsoft.com/en-us/sql/relational-databases/security/password-policy?view=sql-server-ver16#password-complexity). Note: By starting a Docker container with the environment variable ACCEPT_EULA=Y you agree to the terms of Microsoft EULA, see also [MICROSOFT SOFTWARE LICENSE TERMS MICROSOFT SQL SERVER 2019 STANDARD(EN_US)](https://www.microsoft.com/en-us/Useterms/Retail/SQLServerStandard/2019/Useterms_Retail_SQLServerStandard_2019_English.htm).

File: en_US/data-integration/data-bridge-sqlserver.md [hunk 8]
```diff
@@ -0,0 +1,337 @@
...
+```
+
+2. Access the container.
+
+```bash
+docker exec -it sqlserver bash
+```
+
+3. Enter the preset password to connect to the server in the container.
+   - `mssql-tools` has been installed in the SQL Server container provided by Microsoft, but the executable file is not in `$PATH`. You need to specify the executable file path for `mssql-tools`. For more usage information on `mssql-tools`, refer to related documents provided by Microsoft: [sqlcmd-utility](https://learn.microsoft.com/en-us/sql/tools/sqlcmd/sqlcmd-utility?view=sql-server-ver16).
```
Comment (LenaLenaPan) [inline en_US/data-integration/data-bridge-sqlserver.md]:
```suggestion
   - The `mssql-tools` have already been installed in the SQL Server container provided by Microsoft, but the executable file is not in the `$PATH`. Therefore, you need to specify the executable file path for `mssql-tools `before proceeding. For more information on how to use `mssql-tools`, see [sqlcmd-utility](https://learn.microsoft.com/en-us/sql/tools/sqlcmd/sqlcmd-utility?view=sql-server-ver16).
```

File: zh_CN/data-integration/data-bridge-sqlserver.md [hunk 9]
```diff
@@ -0,0 +1,330 @@
...
+- [安装 Microsoft ODBC Driver for SQL Server (macOS)](https://learn.microsoft.com/zh-cn/sql/connect/odbc/linux-mac/install-microsoft-odbc-driver-sql-server-macos?view=sql-server-ver16)
+
+受限于 [Microsoft EULA 条款](https://www.microsoft.com/en-us/Useterms/Retail/SQLServerStandard/2019/Useterms_Retail_SQLServerStandard_2019_ChineseSimplified.htm)，EMQX 提供的 Docker 镜像不带有 msodbcsql17 驱动程序，如需在 Docker 或 Kubernetes 中使用该驱动程序，您需要基于 [EMQX-Enterprise](https://hub.docker.com/r/emqx/emqx-enterprise) 提供的镜像构建带有 ODBC 驱动程序的新镜像以便在连接 SQL Server 数据库时使用 msodbcsql17 驱动程序。使用构建的新镜像，代表您同意 Microsoft SQL Server EULA。
+
+1. 在 EMQX 的仓库中找到对应的 [Dockerfile](https://github.com/emqx/emqx/blob/master/deploy/docker/Dockerfile.msodbc)。您可以将该文件保存至本地。
+
+   下面的示例中 Dockerfile 中的镜像版本为 `emqx/emqx-enterprise:5.0.3-alpha.2`，对于 EMQX-Enterprise 5.0.3 以后的版本，可以根据您需要的 EMQX-Enterprise 版本构建镜像，也可以使用 EMQX-Enterprise 最新版本镜像 `emqx/emqx-enterprise:latest` 进行构建。
+
+   ```bash
```
Comment (JimMoen) [inline zh_CN/data-integration/data-bridge-sqlserver.md]: > \`\`\`dockerfile

File: en_US/data-integration/data-bridge-sqlserver.md [hunk 10]
```diff
@@ -0,0 +1,337 @@
...
+
+Restricted by [Microsoft EULA terms](https://www.microsoft.com/en-us/Useterms/Retail/SQLServerStandard/2019/Useterms_Retail_SQLServerStandard_2019_English.htm), the Docker image provided by EMQX does not include the msodbcsql17 driver. To use it in Docker or Kubernetes, you need to create a new image installed with ODBC driver based on the image provided by [EMQX-Enterprise](https://hub.docker.com/r/emqx/emqx-enterprise) to access the SQL Server database. Using the new image means that you agree to the [Microsoft SQL Server EULA](https://www.microsoft.com/en-us/Useterms/Retail/SQLServerStandard/2019/Useterms_Retail_SQLServerStandard_2019_English.htm).
+
+Follow the instructions below to build a new image:
+
+1. Get the corresponding EMQX [Dockerfile](https://github.com/emqx/emqx/blob/master/deploy/docker/Dockerfile.msodbc). You can save the file locally.
+
+   The image version in this example is `emqx/emqx-enterprise:5.0.3-alpha.2`. You can build the image based on the EMQX-Enterprise version you need, or use the latest version image `emqx/emqx-enterprise:latest`.
+
+   ```bash
```
Comment (JimMoen) [inline en_US/data-integration/data-bridge-sqlserver.md]: > \`\`\`dockerfile

File: en_US/data-integration/data-bridge-sqlserver.md [hunk 11]
```diff
@@ -0,0 +1,337 @@
+# Ingest Data into Microsoft SQL Server
+
+EMQX supports integration with Microsoft SQL Server. You can save client messages and events to Microsoft SQL Server, or record the online status or online/offline of clients by using events to trigger the data update or removal.
+
+{% emqxee %}
+
+::: tip
+
+The data integration with Microsoft SQL Server is currently only supported in EMQX Enterprise 5.0.3 and above.
```
Comment (JimMoen) [inline en_US/data-integration/data-bridge-sqlserver.md]: The data integration with Microsoft SQL Server is only supported in EMQX Enterprise 5.0.3 and above.

File: en_US/data-integration/data-bridge-sqlserver.md [hunk 12]
```diff
@@ -0,0 +1,342 @@
...
+   $ /opt/mssql-tools/bin/sqlcmd -S 127.0.0.1 -U sa
+   $ Password:
+   1>
+   ```
+   
+   ::: tip
+   
+   `mssql-tools` has been installed in the Microsoft SQL Server container provided by Microsoft, but the executable file is not in `$PATH`. You need to specify the executable file path for `mssql-tools`. As for the Docker deployment in this example, the file path should be `opt`.
+   
+   For more usage information on `mssql-tools`, refer to related documents provided by Microsoft: [sqlcmd-utility](https://learn.microsoft.com/en-us/sql/tools/sqlcmd/sqlcmd-utility?view=sql-server-ver16).
```
Comment (LenaLenaPan) [inline en_US/data-integration/data-bridge-sqlserver.md]:
```suggestion
   For more usage information on `mssql-tools`, see [sqlcmd-utility](https://learn.microsoft.com/en-us/sql/tools/sqlcmd/sqlcmd-utility?view=sql-server-ver16). 
```

File: zh_CN/data-integration/data-bridge-sqlserver.md [hunk 13]
```diff
@@ -0,0 +1,336 @@
...
+   $ /opt/mssql-tools/bin/sqlcmd -S 127.0.0.1 -U sa
+   $ Password:
+   1>
+   ```
+   
+   ::: tip
+   
+   Microsoft 提供的 Microsoft SQL Server 容器内已安装 `mssql-tools`，但可执行文件并不在 `$PATH` 中，因此您需要指定可执行文件路径。在上述连接示例中，可执行文件路径为 `opt`。
+   
+   关于更多 `mssql-tools` 的使用，请参与 Microsoft 提供的相关文档： [sqlcmd 实用工具](https://learn.microsoft.com/zh-cn/sql/tools/sqlcmd/sqlcmd-utility?view=sql-server-ver16)。
```
Comment (LenaLenaPan) [inline zh_CN/data-integration/data-bridge-sqlserver.md]:
```suggestion
   关于更多 `mssql-tools` 的使用，请阅读 [sqlcmd 实用工具](https://learn.microsoft.com/zh-cn/sql/tools/sqlcmd/sqlcmd-utility?view=sql-server-ver16)。
```

File: zh_CN/data-integration/data-bridge-sqlserver.md [hunk 14]
```diff
@@ -247,11 +251,14 @@ FileUsage   = 1
...
-8. 在点击 **创建** 按钮完成数据桥接创建之前，您可以使用 **测试连接** 来测试当前 EMQX 到 SQL Server 的连接是否成功。
-9. 点击创建按钮完成数据桥接创建。
 
-至此您已经完成数据桥接创建，SQL Server 数据桥接应该出现在数据桥接列表（**数据集成** -> **数据桥接**）中，**资源状态**为**已连接**。
-接下来将继续创建一条规则来指定需要写入的数据。
+8. 在点击 **创建** 按钮完成数据桥接创建之前，您可以使用 **测试连接** 来测试当前 EMQX 到 Microsoft SQL Server 的连接是否成功。
+
+9. 点击**创建**按钮完成数据桥接创建。
+
+   在弹出的**创建成功**对话框中您可以点击**创建规则**，继续创建规则以指定需要写入 RocketMQ 的数据。您也可以按照[创建 Microsoft SQL Server 数据桥接规则](#创建-microsoft-sql-server-数据桥接规则)章节的步骤来创建规则。
```
Comment (JimMoen) [inline zh_CN/data-integration/data-bridge-sqlserver.md]: Microsoft SQL Server

File: zh_CN/data-integration/data-bridge-sqlserver.md [hunk 15]
```diff
@@ -263,19 +270,18 @@ FileUsage   = 1
...
+        *
+      FROM
+        "t/#"
      ```
-
-   - 如需实现设备上下线状态记录，输入以下 SQL 语法：
-
+   
+   - 如需实现设备上下线状态记录，输入以下 SQL 语句：
+
```
Comment (JimMoen) [inline zh_CN/data-integration/data-bridge-sqlserver.md]: nit: please rm unnessary indent and whitespace.

File: zh_CN/data-integration/data-bridge-sqlserver.md [hunk 16]
```diff
@@ -0,0 +1,336 @@
...
+本节介绍了如何在 EMQX Dashboard 上创建 Microsoft SQL Server 数据桥接以实现对客户端发布消息的存储或设备状态的记录。
+
+1. 在 Dashboard 点击 **数据集成** -> **数据桥接**。
+
+2. 点击页面右上角的**创建**。
+
+3. 在数据桥接类型中选择 Microsoft SQL Server，点击**下一步**。
+
+4. 输入数据桥接名称，要求是大小写英文字母或数字组合。
```
Comment (LenaLenaPan) [inline zh_CN/data-integration/data-bridge-sqlserver.md]:
```suggestion
4. 输入数据桥接名称，要求是大小写英文字母和数字的组合。
```

File: zh_CN/data-integration/data-bridge-sqlserver.md [hunk 17]
```diff
@@ -0,0 +1,336 @@
...
+5. 输入 Microsoft SQL Server 连接信息。
+   - **服务器地址**： `127.0.0.1:1433`，或使用实际的 Microsoft SQL Server 地址和端口
+   - **数据库名字**： `mqtt`
+   - **用户名**： `sa`
+   - **密码**： `mqtt_public1`
+   - **SQL Server Driver 名称**： `ms-sql`，即您在 `odbcinst.ini` 中配置的 DSN Name
+
+6. 根据业务实现需要配置 SQL 模板：
+
+   - 如需实现对指定主题消息的转发，使用如下 SQL 语句完成数据插入。此处为[预处理 SQL](./data-bridges.md#sql-预处理)，字段不应当包含引号，SQL 末尾不要带分号 `;`:
```
Comment (LenaLenaPan) [inline zh_CN/data-integration/data-bridge-sqlserver.md]:
```suggestion
   - 如需实现对指定主题消息的转发，使用如下 SQL 语句完成数据插入。此处为[预处理 SQL](./data-bridges.md#sql-预处理)，字段不应当包含引号，SQL 末尾不要带分号 `;`
```

File: zh_CN/data-integration/data-bridge-sqlserver.md [hunk 18]
```diff
@@ -0,0 +1,336 @@
...
+
+     ```sql
+     insert into t_mqtt_msg(msgid, topic, qos, payload) values ( ${id}, ${topic}, ${qos}, ${payload} )
+     ```
+
+   - 如需实现实现设备上下线状态记录，使用如下 SQL 语句完成数据插入：
```
Comment (LenaLenaPan) [inline zh_CN/data-integration/data-bridge-sqlserver.md]:
```suggestion
   - 如需实现实现设备上下线状态记录，可使用如下 SQL 语句完成数据插入：
```

File: zh_CN/data-integration/data-bridge-sqlserver.md [hunk 19]
```diff
@@ -0,0 +1,336 @@
...
+
+8. 在点击 **创建** 按钮完成数据桥接创建之前，您可以使用 **测试连接** 来测试当前 EMQX 到 Microsoft SQL Server 的连接是否成功。
+
+9. 点击**创建**按钮完成数据桥接创建。
+
+   在弹出的**创建成功**对话框中您可以点击**创建规则**，继续创建规则以指定需要写入 Microsoft SQL Server 的数据。您也可以按照[创建 Microsoft SQL Server 数据桥接规则](#创建-microsoft-sql-server-数据桥接规则)章节的步骤来创建规则。
+
+至此您已经完成数据桥接创建，Microsoft SQL Server 数据桥接应该出现在数据桥接列表（**数据集成** -> **数据桥接**）中，**资源状态**为**已连接**。
+
+### 创建 Microsoft SQL Server 数据桥接规则
```
Comment (LenaLenaPan) [inline zh_CN/data-integration/data-bridge-sqlserver.md]:
```suggestion
### 创建数据转发规则
```

File: zh_CN/data-integration/data-bridge-sqlserver.md [hunk 20]
```diff
@@ -0,0 +1,336 @@
...
+       "$events/client_connected", "$events/client_disconnected"
+     ```
+
+4. 点击**添加动作**，在动作下拉框中选择**使用数据桥接转发**选项，选择先前创建好的 Microsoft SQL Server 数据桥接。
+
+5. 点击最下方**创建**按钮完成规则创建。
+
+至此您已经完成整个创建过程，可以前往 **数据集成** -> **Flows** 页面查看拓扑图，此时应当看到 `t/#` 主题的消息经过名为 `my_rule` 的规则处理，处理结果交由 Microsoft SQL Server 存储。
+
+### 测试连接和规则
```
Comment (LenaLenaPan) [inline zh_CN/data-integration/data-bridge-sqlserver.md]:
```suggestion
### 测试数据桥接和规则
```

**Review by LenaLenaPan — COMMENTED** [review]
统一中文标题

**Review by wivwiv — APPROVED** [review]

**Comment by LenaLenaPan** [comment]
Uuse pr #1865 to solve the conflict

**Pull request closed without merging**